Library/Source/NativeEngine.cpp

```diff
@@ -151,6 +151,65 @@ namespace babylon
             }
         }
 
+        void GenerateMipmap(uint8_t* __restrict source, int size, int channels, uint8_t* __restrict dest)
+        {
+            int mipsize = size / 2;
+
+            for (int y = 0; y < mipsize; ++y)
+            {
+                for (int x = 0; x < mipsize; ++x)
+                {
+                    for (int c = 0; c < channels; ++c)
+                    {
+                        int index = channels * ((y * 2) * size + (x * 2)) + c;
+                        int sum_value = 4 >> 1;
+                        sum_value += source[index + channels * (0 * size + 0)];
+                        sum_value += source[index + channels * (0 * size + 1)];
+                        sum_value += source[index + channels * (1 * size + 0)];
+                        sum_value += source[index + channels * (1 * size + 1)];
+                        dest[channels * (y * mipsize + x) + c] = (uint8_t)(sum_value / 4);
+                    }
+                }
+            }
+        }
+
+        const bgfx::Memory* GenerateMipMaps(const bimg::ImageContainer& image)
+        {
+            auto imageDataRef{ bgfx::makeRef(image.m_data, image.m_size) };
+
+            bool widthIsPowerOf2 = ((image.m_width & (~image.m_width + 1)) == image.m_width);
+            bool supportedFormat = image.m_format == bimg::TextureFormat::RGB8 || image.m_format == bimg::TextureFormat::RGBA8 || image.m_format == bimg::TextureFormat::BGRA8;
+            if (image.m_width == image.m_height && image.m_width && widthIsPowerOf2 && supportedFormat)
+            {
+                auto channelCount = (image.m_format == bimg::TextureFormat::RGB8) ? 3 : 4;
+                unsigned int mipMapCount = static_cast<unsigned int>(log2(image.m_width));
```

Comment:
do you need to cast here? looks like image.m_width is a `uint32_t`? maybe use `uint32_t` instead of `unsigned int`?

Reply:
log2 returns float or double

Reply:
Oops, yeah, forgot about the math part. lol